File: Ejercicios_etapa_1/Lua/Ejercicio_6/prueba.test.lua
-- prueba.test.lua
package.path = "./?.lua;" .. package.path
local metaTablas = require("metaTablas")

describe("accederArray", function()
  it("debe retornar el elemento en la posición indicada", function()
    local arr = {'a', 'b', 'c'}
    -- En Lua, siendo 1-indexado, el índice 2 corresponde al segundo elemento.
    assert.are.equal('b', metaTablas.accederArray(arr, 2))
  end)
end)

describe("manipulaArray", function()
  it("debe retornar el segundo elemento del array creado", function()
    -- Se espera que se inserten tres elementos y se retorne el elemento en la posición 2.
    assert.are.equal(2, metaTablas.manipulaArray({}, 1, 2, 3))
  end)
end)

describe("sumarElementos", function()
  it("debe retornar la suma de todos los elementos cuando son números", function()
    assert.are.equal(420, metaTablas.sumarElementos({200, 200, 20}))
  end)
  it("debe retornar 'Datos inválidos' si algún elemento no es un número", function()
    assert.are.equal("Datos inválidos", metaTablas.sumarElementos({200, "200", 20}))
  end)
end)

describe("generarRango", function()
  it("debe retornar una tabla con números desde 0 hasta n (inclusive)", function()
    local esperado = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10}
    assert.are.same(esperado, metaTablas.generarRango(10))
  end)
  it("debe retornar 'Datos inválidos' si n no es un número", function()
    assert.are.equal("Datos inválidos", metaTablas.generarRango("a"))
  end)
end)

describe("organizarPorTipo", function()
  it("debe devolver una tabla con 'strings' y 'numbers' separados", function()
    local resultado = metaTablas.organizarPorTipo({1, "hola", 3, "mundo"})
    local esperado = {
      numbers = {1, 3},
      strings = {"hola", "mundo"}
    }
    assert.are.same(esperado, resultado)
  end)
end)

describe("miMapa", function()
  it("debe retornar un nuevo array con la función aplicada a cada elemento", function()
    local doble = function(x) return x * 2 end
    local resultado = metaTablas.miMapa({1, 2, 3, 4}, doble)
    local esperado = {2, 4, 6, 8}
    assert.are.same(esperado, resultado)
  end)
end)

describe("miFilter", function()
  it("debe retornar un nuevo array con los elementos que cumplen la condición", function()
    local esPar = function(num) return num % 2 == 0 end
    local resultado = metaTablas.miFilter({1, 2, 3, 4, 5, 6}, esPar)
    local esperado = {2, 4, 6}
    assert.are.same(esperado, resultado)
  end)
end)

describe("miReduce", function()
  it("debe retornar el resultado acumulado tras aplicar la función reductora", function()
    local sumar = function(acumulador, valor) return acumulador + valor end
    assert.are.equal(10, metaTablas.miReduce({1, 2, 3, 4}, sumar, 0))
  end)
end)

describe("usaMap", function()
  it("debe retornar un nuevo array usando el método simulado de map", function()
    local multiplicarPorTres = function(x) return x * 3 end
    local resultado = metaTablas.usaMap({1, 2, 3, 4}, multiplicarPorTres)
    local esperado = {3, 6, 9, 12}
    assert.are.same(esperado, resultado)
  end)
end)

describe("usaFilter", function()
  it("debe retornar un nuevo array usando el método simulado de filter", function()
    local esImparejo = function(num) return num % 2 ~= 0 end
    local resultado = metaTablas.usaFilter({1, 2, 3, 4, 5, 6}, esImparejo)
    local esperado = {1, 3, 5}
    assert.are.same(esperado, resultado)
  end)
end)

describe("usaReduce", function()
  it("debe retornar el resultado acumulado usando el método simulado de reduce", function()
    local multiplicar = function(ac, val) return ac * val end
    assert.are.equal(24, metaTablas.usaReduce({1, 2, 3, 4}, multiplicar, 1))
  end)
end)

File: Ejercicios_etapa_1/00_Javascript/Ejercicio_6/prueba.test.js
import { describe, it, expect } from 'vitest'
import { 
  accederArray, 
  manipulaArray, 
  sumarElementos, 
  generarRango, 
  organizarPorTipo, 
  miMapa, 
  miFilter, 
  miReduce, 
  usaMap, 
  usaFilter, 
  usaReduce 
} from './index.js'

describe("accederArray", () => {
  it("debe retornar el elemento en el array según el índice", () => {
    const arr = ['a', 2, 'c']
    expect(accederArray(arr, 0)).toBe('a')
    expect(accederArray(arr, 1)).toBe(2)
    expect(accederArray(arr, 2)).toBe('c')
  })
})

describe("manipulaArray", () => {
  it("debe retornar el segundo elemento del nuevo array", () => {
    // For example, if the function pushes "x", "y", "z", then the second element is "y"
    expect(manipulaArray([], "x", "y", "z")).toBe("y")
  })
})

describe("sumarElementos", () => {
  it("debe retornar la suma de todos los elementos si son números", () => {
    expect(sumarElementos([200, 200, 20])).toBe(420)
  })
  it("debe retornar 'Datos inválidos' si algún elemento no es un número", () => {
    expect(sumarElementos([200, "200", 20])).toBe("Datos inválidos")
  })
})

describe("generarRango", () => {
  it("debe retornar un array con números desde 0 hasta n (inclusive)", () => {
    expect(generarRango(5)).toEqual([0, 1, 2, 3, 4, 5])
  })
  it("debe retornar 'Datos inválidos' si n no es un número", () => {
    expect(generarRango("a")).toBe("Datos inválidos")
  })
})

describe("organizarPorTipo", () => {
  it("debe devolver un objeto con las propiedades 'strings' y 'numbers'", () => {
    const input = [1, "hola", 3, "mundo", true]
    // Suponiendo que los booleans se ignoran, se espera:
    expect(organizarPorTipo(input)).toEqual({
      numbers: [1, 3],
      strings: ["hola", "mundo"]
    })
  })
})

describe("miMapa", () => {
  it("debe retornar un nuevo array con la función aplicada a cada elemento", () => {
    const doble = x => x * 2
    expect(miMapa([1, 2, 3, 4], doble)).toEqual([2, 4, 6, 8])
  })
})

describe("miFilter", () => {
  it("debe retornar un nuevo array con los elementos que cumplen la condición", () => {
    const esPar = x => x % 2 === 0
    expect(miFilter([1, 2, 3, 4, 5, 6], esPar)).toEqual([2, 4, 6])
  })
})

describe("miReduce", () => {
  it("debe retornar el resultado acumulado tras aplicar la función reductora", () => {
    const sumar = (acc, val) => acc + val
    expect(miReduce([1, 2, 3, 4], sumar, 0)).toBe(10)
  })
})

describe("usaMap", () => {
  it("debe retornar un nuevo array usando el método .map", () => {
    const multiplicarPorTres = x => x * 3
    expect(usaMap([1, 2, 3, 4], multiplicarPorTres)).toEqual([3, 6, 9, 12])
  })
})

describe("usaFilter", () => {
  it("debe retornar un nuevo array usando el método .filter", () => {
    const esImparejo = x => x % 2 !== 0
    expect(usaFilter([1, 2, 3, 4, 5, 6], esImparejo)).toEqual([1, 3, 5])
  })
})

describe("usaReduce", () => {
  it("debe retornar el resultado acumulado usando el método .reduce", () => {
    const multiplicar = (acc, val) => acc * val
    expect(usaReduce([1, 2, 3, 4], multiplicar, 1)).toBe(24)
  })
})
